### ISSUE TITLE
ssh key issue when using travis-ci/dpl deis deploy

### ISSUE THREAD

**Issue by ianblenke** [deis/deis#3285]
Successive deploys have left deis in a weird state. During a deploy, we recieved a 500 error from the controller when adding an ssh key. The deis-controller log shows:

```
Mar 12 21:15:08 ip-10-113-1-220.ec2.internal sh[8984]: 10.113.0.135 "POST /v1/auth/login/ HTTP/1.0" 200 - "python-requests/2.5.1 CPython/2.7.6 Linux/3.13.0-46-generic"
Mar 12 21:15:09 ip-10-113-1-220.ec2.internal sh[8984]: 172.17.42.1 "GET /v1/apps/mother HTTP/1.0" 200 - "python-requests/2.5.1 CPython/2.7.6 Linux/3.13.0-46-generic"
Mar 12 21:15:09 ip-10-113-1-220.ec2.internal sh[8984]: 10.113.2.108 "GET /v1/apps/mother/containers HTTP/1.0" 200 - "python-requests/2.5.1 CPython/2.7.6 Linux/3.13.0-46-generic"
Mar 12 21:15:09 ip-10-113-1-220.ec2.internal sh[8984]: 10.113.0.135 "GET /v1/apps/mother/domains HTTP/1.0" 200 - "python-requests/2.5.1 CPython/2.7.6 Linux/3.13.0-46-generic"
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: ERROR Internal Server Error: /v1/keys
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: Traceback (most recent call last):
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/base.py", line 112, in get_response
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: response = wrapped_callback(request, *callback_args, **callback_kwargs)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/django/views/decorators/csrf.py", line 57, in wrapped_view
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: return view_func(*args, **kwargs)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/rest_framework/viewsets.py", line 85, in view
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: return self.dispatch(request, *args, **kwargs)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/rest_framework/views.py", line 407, in dispatch
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: response = self.handle_exception(exc)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/rest_framework/views.py", line 404, in dispatch
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: response = handler(request, *args, **kwargs)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/app/api/views.py", line 62, in create
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: return super(BaseDeisViewSet, self).create(request, *args, **kwargs)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/rest_framework/mixins.py", line 21, in create
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: self.perform_create(serializer)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/app/api/viewsets.py", line 20, in perform_create
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: obj = serializer.save(owner=self.request.user)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/rest_framework/serializers.py", line 164, in save
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: self.instance = self.create(validated_data)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/rest_framework/serializers.py", line 766, in create
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: instance = ModelClass.objects.create(**validated_data)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/django/db/models/manager.py", line 157, in create
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: return self.get_queryset().create(**kwargs)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/django/db/models/query.py", line 322, in create
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: obj.save(force_insert=True, using=self.db)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/django/db/models/base.py", line 545, in save
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: force_update=force_update, update_fields=update_fields)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/django/db/models/base.py", line 573, in save_base
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/django/db/models/base.py", line 654, in _save_table
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/django/db/models/base.py", line 687, in _do_insert
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: using=using, raw=raw)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/django/db/models/manager.py", line 232, in _insert
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: return insert_query(self.model, objs, fields, **kwargs)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/django/db/models/query.py", line 1514, in insert_query
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: return query.get_compiler(using=using).execute_sql(return_id)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/django/db/models/sql/compiler.py", line 903, in execute_sql
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: cursor.execute(sql, params)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/django/db/backends/util.py", line 53, in execute
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: return self.cursor.execute(sql, params)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/django/db/utils.py", line 99, in __exit__
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: six.reraise(dj_exc_type, dj_exc_value, traceback)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: File "/usr/local/lib/python2.7/dist-packages/django/db/backends/util.py", line 53, in execute
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: return self.cursor.execute(sql, params)
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: IntegrityError: duplicate key value violates unique constraint "api_key_owner_id_4c18718c256f9c15_uniq"
Mar 12 21:15:10 ip-10-113-1-220.ec2.internal sh[8984]: DETAIL:  Key (owner_id, id)=(3, testing-te-worker-test1-373-linux-1) already exists.
```

When I try and delete the key, I see:

```
iblenke@ip-10-113-0-135 ~ $ deis login https://deis.mydomain.com
username: travis
password:
Logged in as travis
iblenke@ip-10-113-0-135 ~ $ deis keys
=== travis Keys
testing-te-worker-test1-373-linux-1 ssh-rsa AAAAB3Nz...y8vs3g9DnD
iblenke@ip-10-113-0-135 ~ $ deis keys help
Usage: deis keys:list
iblenke@ip-10-113-0-135 ~ $ deis keys:remove testing-te-worker-test1-373-linux-1
Removing testing-te-worker-test1-373-linux-1 SSH Key... 403 FORBIDDEN
Detail:
You do not have permission to perform this action.
```

So as an admin, I begrudgingly make this user an admin:

```shell
$ deis perms:create travis --admin
Adding travis to system administrators... done
```

And I can now delete this key as the user:

```shell
$ deis keys:remove testing-te-worker-test1-373-linux-1
Removing testing-te-worker-test1-373-linux-1 SSH Key... done
```

I'm going to bet we don't have this permissions problem again now that the user is an admin.

**Comment by bacongobbler** [comment]
If you inspect the database, can you confirm that the `travis` user is indeed ID 3?

**Comment by carmstrong** [comment]
@ianblenke Any more investigation on this?

**Comment by ianblenke** [comment]
@bacongobbler, the `travis` user is indeed ID 3.

```
root@dc48864fc1b1:/app# ./manage.py shell
...
>>> from django.contrib.auth.models import User
>>> u = User.objects.get(username__exact='travis')
>>> print u.id
3
```

It also presently has 3 keys defined:

```
$ deis keys
=== travis Keys
testing-te-worker-test1-373-linux-2 ssh-rsa AAAAB3Nz...14rWqPxzV1
testing-te-worker-test1-373-linux-1 ssh-rsa AAAAB3Nz...+Pj/e863OB
testing-te-worker-test1-18934-linux-2 ssh-rsa AAAAB3Nz...DfpPOVy2Vl
```

As mentioned earlier in this issue, that `testing-te-worker-test1-373-linux-1` mention above is different than the one it used before.

I'm currently trying to troubleshoot why it isn't working once more all of the sudden. The travis user is not an admin, suggesting our database has reverted again. I'm not getting that traceback as before, it's not getting that far: for some reason the deis login isn't working for travis automatically, but I _can_ login as the travis user manually. Anyway, I'm fighting with the dpl stuff now adding debugging to understand what it is doing.

**Comment by eMxyzptlk** [comment]
@ianblenke I'm having the same problem, I cannot delete the keys unless my user is an administrator. Also, I am experiencing issues deploying from travis and I believe we're both hitting travis-ci/dpl#240

**Comment by eMxyzptlk** [comment]
@ianblenke Workaround the travis issue: travis-ci/dpl#240

Basically you need `$HOME/.local/bin` to be in the PATH.

**Comment by jorihardman** [comment]
I'm still have this issue in v1.6.1. I get a 403 when trying to remove keys with a non-admin user.

**Comment by jorihardman** [comment]
@mboersma Still having the issue in v1.7.3

Non-admin users cannot remove keys they upload.

**Comment by bacongobbler** [comment]
this was recently fixed and will get in for v1.9: https://github.com/deis/deis/pull/3931
